ark/jsonschema/json.ts

```diff
@@ -0,0 +1,133 @@
+> = t extends never ? never : t & inferJsonSchema<Omit<schema, kind>>
+
+type inferJsonSchemaTypeNoKeywords<
+	schema extends JsonSchema.TypeWithNoKeywords,
+	t
+> =
+	schema["type"] extends "boolean" ? t & boolean
+	: schema["type"] extends "null" ? t & null
+	: never
+
+export type inferJsonSchema<schema, t = unknown> =
+	schema extends true ? JsonSchema.Json
+	: schema extends false ? never
+	: schema extends Record<PropertyKey, never> ? JsonSchema.Json
+	: schema extends array ? inferJsonSchema<schema[number], t>
+	: schema extends JsonSchema.CompositionKeywords ?
+		inferJsonSchemaComposition<schema, t>
+	: schema extends JsonSchemaConst<infer c> ?
+		inferJsonSchemaConstraint<schema, t & c, "const">
+	: schema extends JsonSchemaEnum<infer e> ?
+		inferJsonSchemaConstraint<schema, t & e, "enum">
+	: schema extends JsonSchema.TypeWithNoKeywords ?
+		inferJsonSchemaTypeNoKeywords<schema, t>
+	: schema extends JsonSchema.ArraySchema ? inferJsonSchemaArray<schema>
+	: schema extends JsonSchema.NumberSchema ? t & inferJsonSchemaNumber<schema>
+	: schema extends JsonSchema.ObjectSchema ? t & inferJsonSchemaObject<schema>
+	: schema extends JsonSchema.StringSchema ? t & inferJsonSchemaString<schema>
+	: t extends {} ? t
+	: ErrorMessage<"Failed to infer JSON Schema">
+
+export const innerParseJsonSchema: Type<
+	(In: JsonSchema.Schema) => Out<Type<unknown, any>>
+> = JsonSchema.Schema.pipe(
+	(jsonSchema: JsonSchema.Schema): Type<unknown, any> => {
+		if (typeof jsonSchema === "boolean") {
+			if (jsonSchema) return JsonSchema.Json
+			else return type("never") // No runtime value ever passes validation for JSON schema of 'false'
+		}
+
+		if (Array.isArray(jsonSchema)) {
+			return (
+				parseJsonSchemaCompositionKeywords({ anyOf: jsonSchema }) ??
+				throwParseError(
+					"Failed to convert root array of JSON Schemas to an anyOf schema"
+				)
+			)
+		}
+
+		const constAndOrEnumValidator = parseJsonSchemaAnyKeywords(jsonSchema)
+		const compositionValidator = parseJsonSchemaCompositionKeywords(jsonSchema)
+
+		const preTypeValidator: Type | undefined =
+			constAndOrEnumValidator ?
+				compositionValidator ? compositionValidator.and(constAndOrEnumValidator)
+				:	constAndOrEnumValidator
+			:	compositionValidator
+
+		if ("type" in jsonSchema) {
+			let typeValidator: Type
+			switch (jsonSchema.type) {
+				case "array":
+					typeValidator = validateJsonSchemaArray.assert(jsonSchema)
+					break
+				case "boolean":
+				case "null":
+					typeValidator = type(jsonSchema.type)
+					break
+				case "integer":
+				case "number":
+					typeValidator = validateJsonSchemaNumber.assert(jsonSchema)
+					break
+				case "object":
+					typeValidator = validateJsonSchemaObject.assert(jsonSchema)
+					break
+				case "string":
+					typeValidator = validateJsonSchemaString.assert(jsonSchema)
+					break
+				default:
+					throwParseError(
+						// @ts-expect-error -- All valid 'type' values should be handled above
```

Comment:
`@ts-expect-error` should (almost) never be used in source.

In a case like this, I'd use `jsonSchema.type satisfies never`

Reply:
I'm not quite sure what you mean by this. I'm still getting an error, but think that might just be me misunderstanding how to use this?
![image](https://github.com/user-attachments/assets/80ba274f-9cc5-41e5-b18a-65e2f8644ca4)